### PR TITLE
fix failed validation on rake db:seed

### PR DESCRIPTION
**Pull request by ethanluoyc**
Fixes https://www.pivotaltracker.com/story/show/113736889

### CODE REVIEW

**Comment by tansaku** [comment]
thanks @ethanluoyc - great start!

I think we might need to test drive this bug fix in order to get the CI to pass :-)

Will you need help approaching that?

**Comment by tansaku** [comment]
also note you have "[Fixes #113736889]" in the description, when the pivotal integration needs it in the github commit ...

**Comment by ethanluoyc** [comment]
Ok, I will also complete the test driven part since I have already started on this.
For multiple commits do I need to merge them into one before you accept the PR?

**Comment by ethanluoyc** [comment]
> I think we might need to test drive this bug fix in order to get the CI to pass :-)

Where can I see the failed builds? coveralls says decreased but I didn't find any errors in it.

**Comment by tansaku** [comment]
@ethanluoyc regarding the 'Fix' in the commit - apologies I didn't see it due to not clicking on the see more bit.  Having it in one single commit is just fine.  No need to have it in all, although usually we might put it in the last commit.

We currently don't require merging all the commits into one, although we have worked with other projects where that is a requirement.

We are a learning project so I think the ability to look back at a pull request and see the different components is valuable.  It also can be tricky to rebase commits, and it also makes the conversation history in a pull request get out of order.

The build has not technically "failed" in this instance, so much as you added code without adding a test so the "test coverage" decreased.  If you add a test that tests the code the coverage will increase or stay the same and the build will pass.  Make sense?

**Comment by ethanluoyc** [comment]
@tansaku feedback suggests that the newly included code is not covered. 
I want to ensure that CreateOrganisationFromArray can be tested independent of the Organisation model , but current method prohibits me from doing so.

**Comment by tansaku** [comment]
@ethanluoyc sounds like some refactoring is in order - do you want me to write some example code? Or perhaps we can work on this together?

**Comment by ethanluoyc** [comment]
Yes that will be appreciated :smile:

**Comment by tansaku** [comment]
@ethanluoyc I'm available for a hangout now if you are free ...

File: app/models/organisation.rb [hunk 1]
```diff
@@ -89,6 +89,10 @@ def self.create_and_validate(attributes)
     create!(attributes.select{|k,v| !v.nil?})
   end
 
+  def self.create_and_substitute_with_empty(attributes)
```
Comment (tansaku) [inline app/models/organisation.rb]: there's no matching unit test for this method

File: app/services/create_organisation_from_array.rb [hunk 2]
```diff
@@ -15,13 +15,24 @@ def initialize(organisation_repository, row)
...
+    else
+      @organisation_repository.create_and_substitute_with_empty(
+          name: @csv_organisation.organisation_name,
+          address: @csv_organisation.capitalize_address,
+          description: @csv_organisation.description,
+          postcode: @csv_organisation.postcode,
+          website: @csv_organisation.website,
+          telephone: @csv_organisation.telephone)
+    end
+
```
Comment (tansaku) [inline app/services/create_organisation_from_array.rb]: could refactor like this:

```ruby
  def create(validate)
    return nil if @csv_organisation.is_organisation_removed? 
    return nil if organisation_already_exists?
    create_method = validate ? :create_and_validate : :create_and_substitute_with_empty
    @organisation_repository.send(create_method,
        name: @csv_organisation.organisation_name,
        address: @csv_organisation.capitalize_address,
        description: @csv_organisation.description,
        postcode: @csv_organisation.postcode,
        website: @csv_organisation.website,
        telephone: @csv_organisation.telephone)
  end
```

File: app/models/organisation.rb [hunk 3]
```diff
@@ -89,6 +89,10 @@ def self.create_and_validate(attributes)
...
+    create!(attributes.map { |k, v| attributes[k] = v.nil? ? ' ' : v })
```
Comment (tansaku) [inline app/models/organisation.rb]: note you are turning a hash into an array here, which I don't think is what you want ... see http://stackoverflow.com/questions/5189161/changing-every-value-in-a-hash-in-ruby perhaps

File: spec/models/organisation_spec.rb [hunk 4]
```diff
@@ -291,6 +291,15 @@ def build_org_with_computed_fields_and_updated_at org, updated_at = nil
...
     end
 
+    it 'should be able to substitute with empty string when data is missing' do
+      attributes = {name: '',
+                    address: '',
+                    description: '',
+                    postcode: '',
+                    website: '',
+                    telephone: ''}
+      expect { Organisation.create_and_substitute_with_empty attributes }.not_to raise_error
```
Comment (tansaku) [inline spec/models/organisation_spec.rb]: we could say described_class here ... my miss ... although I have mixed feelings about it's use anyway :-)

**Comment by tansaku** [comment]
@ethanluoyc great work getting this all sorted out, and helping identify our rubocop config was out of date into the bargain.

Since this is blocking others getting setup I'm tempted to merge this in without any further changes, although there are some places we could work on line length, and I start to question many of the methods that are in the organisation class, but we shouldn't get sidetracked on that here - I'm going to create another chore to that effect so it can be worked on separately:

https://www.pivotaltracker.com/story/show/113926599

**Comment by ethanluoyc** [comment]
@tansaku It is my intention, too. Let's merge this then. Rubocop config can be set up in another story.

**Pull request merged**